src/core/server/elasticsearch/version_check/ensure_es_version.ts

```diff
@@ -0,0 +1,160 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * ES and Kibana versions are locked, so Kibana should require that ES has the same version as
+ * that defined in Kibana's package.json.
+ */
+
+import { interval } from 'rxjs';
+import { map, switchMap, distinctUntilChanged, catchError, startWith } from 'rxjs/operators';
+import {
+  esVersionCompatibleWithKibana,
+  esVersionEqualsKibana,
+} from './es_kibana_version_compatability';
+import { Logger } from '../../logging';
+import { APICaller } from '..';
+
+export interface PollEsNodesVersionOptions {
+  callWithInternalUser: APICaller;
+  log: Logger;
+  kibanaVersion: string;
+  ignoreVersionMismatch: boolean;
+  esVersionCheckInterval: number;
+}
+
+interface NodeInfo {
+  version: string;
+  ip: string;
+  http: {
+    publish_address: string;
+  };
+  name: string;
+}
+
+export interface NodesInfo {
+  nodes: {
+    [key: string]: NodeInfo;
+  };
+}
+
+export interface NodesVersionCompatibility {
+  isCompatible: boolean;
+  message?: string;
+  incompatibleNodes: NodeInfo[];
+  warningNodes: NodeInfo[];
+  kibanaVersion: string;
+}
+
+function getHumanizedNodeName(node: NodeInfo) {
+  const publishAddress = node?.http?.publish_address + ' ' || '';
+  return 'v' + node.version + ' @ ' + publishAddress + '(' + node.ip + ')';
+}
+
+export function mapNodesVersionCompatibility(
+  nodesInfo: NodesInfo,
+  kibanaVersion: string,
+  ignoreVersionMismatch: boolean
+): NodesVersionCompatibility {
+  const nodes = Object.keys(nodesInfo.nodes)
+    .sort() // Sorting ensures a stable node ordering for comparison
+    .map(key => nodesInfo.nodes[key])
+    .map(node => Object.assign({}, node, { name: getHumanizedNodeName(node) }));
+
+  // Aggregate incompatible ES nodes.
+  const incompatibleNodes = nodes.filter(
+    node => !esVersionCompatibleWithKibana(node.version, kibanaVersion)
+  );
+
+  // Aggregate ES nodes which should prompt a Kibana upgrade. It's acceptable
+  // if ES and Kibana versions are not the same as long as they are not
+  // incompatible, but we should warn about it.
+  // Ignore version qualifiers https://github.com/elastic/elasticsearch/issues/36859
+  const warningNodes = nodes.filter(node => !esVersionEqualsKibana(node.version, kibanaVersion));
+
+  let message;
+  if (incompatibleNodes.length > 0) {
+    const incompatibleNodeNames = incompatibleNodes.map(node => node.name).join(', ');
+    if (ignoreVersionMismatch) {
+      message = `Ignoring version incompatibility between Kibana v${kibanaVersion} and the following Elasticsearch nodes: ${incompatibleNodeNames}`;
+    } else {
+      message = `This version of Kibana (v${kibanaVersion}) is incompatible with the following Elasticsearch nodes in your cluster: ${incompatibleNodeNames}`;
+    }
+  } else if (warningNodes.length > 0) {
```

Comment:
Is it not possible to have both incompatibleNodes and warningNodes? Example:

ES nodes: 7.1.0, 7.2.0, 7.2.1
Kibana Version: 7.2.0

This is sort of edge that doesn't necessarily need to be fixed, but might be worth commenting that the incompatibleNodes message takes precedence over the warningNodes message.